packages/webpack/src/buildtime/modulesData.js

```diff
@@ -0,0 +1,242 @@
+
+/**
+ * @import WebPack from 'webpack'
+ */
+
+/**
+ * @remarks
+ * Webpack has a concept of ignored modules. When a module is ignored, a
+ * carve-out is necessary in policy enforcement for it because the ID that
+ * webpack creates for it is not exactly helpful. example outcome in the bundle:
+ * `const nodeCrypto = __webpack_require__(/*! crypto *\/ "?0b7d");` Sadly, even
+ * treeshaking doesn't eliminate that module. It's left there and failing to
+ * work when reached by runtime policy enforcement. Below is the most reliable
+ * way I've found to date to identify ignored modules.
+ * @param {WebPack.Module} m
+ * @returns {boolean}
+ */
+const isIgnoredModule = (m) => {
+  return Boolean(
+    m.type === JAVASCRIPT_MODULE_TYPE_DYNAMIC &&
+      // @ts-expect-error BAD TYPES
+      m.identifierStr?.startsWith('ignored')
+  )
+}
+
+/**
+ * Checks if a module is a context module.
+ *
+ * @param {any} m - The module to check.
+ * @param {string} moduleClass - The class of the module.
+ * @returns {boolean} - Returns true if the module is a context module,
+ *   otherwise false.
+ */
+const isContextModule = (m, moduleClass) => moduleClass === 'ContextModule'
+
+/**
+ * Identifies an asset that webpack includes in dist by default without setting
+ * any loaders explicitly.
+ *
+ * @param {WebPack.Module} m
+ * @returns {m is WebPack.NormalModule}
+ */
+const isAmbientAsset = (m) =>
+  m.type === 'asset/resource' &&
+  'resource' in m &&
+  'loaders' in m &&
+  Array.isArray(m.loaders) &&
+  m.loaders.length === 0
+
+/**
+ * @param {WebPack.Module} m
+ * @param {string} moduleClass
+ * @returns {m is WebPack.ExternalModule} // TODO: this is not true anymore, but
+ *   there's no superclass of all reasonable module types
+ */
+const isExternalModule = (m, moduleClass) =>
+  ['ExternalModule'].includes(moduleClass) &&
+  'externalType' in m &&
+  m.externalType !== undefined
+/**
+ * @param {WebPack.Module} m
+ * @param {string} moduleClass
+ * @returns {m is import('./policyGenerator.js').InspectableWebpackModule}
+ */
+const isInspectableModule = (m, moduleClass) =>
+  'userRequest' in m ||
+  m.type?.startsWith('javascript') ||
+  isExternalModule(m, moduleClass)
+
+/**
+ * @typedef {{
+ *   module: WebPack.Module
+ *   moduleId: string | number | null
+ * }} IdentifiedModule
+ */
+
+/**
+ * Creates an analyzer for webpack modules that tracks various module types and
+ * their properties
+ *
+ * @param {Object} opts
+ * @param {WebPack.WebpackError[]} opts.mainCompilationWarnings - Array to store
+ *   compilation warnings
+ * @param {IdentifiedModule[]} opts.allIdentifiedModules
+ * @returns {{
+ *   inspectable: import('./policyGenerator.js').InspectableWebpackModule[]
+ *   contextModules: { moduleId: string | number; context: string }[]
+ *   knownPaths: { path: string; moduleId: string | number }[]
+ *   unenforceableModuleIds: (string | number)[]
+ *   externals: Record<string | number, string>
+ * }}
+ */
+exports.analyzeModules = ({
```

Comment:
(opinion) I would probably not define these types inline since it's rather verbose

(also `@import`)